### ISSUE TITLE
antsBrainExtraction

### ISSUE THREAD

**Issue by pdeman** [stnava/ANTs#425]
Hi,

I am trying to use antsBrainExtraction on T2 only:
`./antsBrainExtraction.sh -d 3 -a /testT2/T2pre_2014-11-25/0001GRE_T2pre_2014-11-25.nii -o /tmp/testT2seg.nii`

I need to add brain extraction template and brain extraction probability mask apparently.
in the help it's said to use for example the LPBA40 dataset but I can't find it.
where could I download it ?

Regards,
Pierre

**Comment by ntustison** [comment]
[Here's](https://figshare.com/articles/ANTs_ANTsR_Brain_Templates/915436) a 5 different templates from this [paper](https://www.ncbi.nlm.nih.gov/pubmed/24879923).

**Comment by pdeman** [comment]
Hi, I have seen that but it says:

>  Each template contains an average T1 neuroimage
and I am trying to segment T2, is that a problem ?

I tried :+1: 
` ./antsBrainExtraction.sh -d 3 -a /testT2/T2pre_2014-11-25/0001GRE_T2pre_2014-11-25.nii -o /tmp/testT2seg.nii -e /MICCAI2012-Multi-Atlas-Challenge-Data/T_template0.nii.gz -m /MICCAI2012-Multi-Atlas-Challenge-Data/T_template0_BrainCerebellumProbabilityMask.nii.gz`
using one of the dataset on your link without sucess.

these templates contain many files and I am not sure to use the good ones.

**Comment by ntustison** [comment]
> ./antsBrainExtraction.sh -d 3 -a /testT2/T2pre_2014-11-25/0001GRE_T2pre_2014-11-25.nii -o /tmp/testT2seg.nii -e /MICCAI2012-Multi-Atlas-Challenge-Data/T_template0.nii.gz -m /MICCAI2012-Multi-Atlas-Challenge-Data/T_template0_BrainCerebellumProbabilityMask.nii.gz
using one of the dataset on your link without sucess.

Perhaps it's something else to do with your data other than a T1/T2 difference.  For example, the data you used has a particular defacing preprocessing step.  If your images don't have this then it wouldn't be a good idea to use this particular data set.

> at the end, is it possible to have the surface grey/white has a .gii file or equivalent ?

You can see if ``antsSurf`` will do what you need.  See ``-o`` option.

**Comment by pdeman** [comment]
how can I get the antsSurf function ?
I have seen this: [https://github.com/stnava/antsSurf](url)
but there isn't the antsSurf function in it.

**Comment by ntustison** [comment]
https://github.com/stnava/ANTs/blob/master/Examples/antsSurf.cxx

You have to enable VTK.

**Comment by pdeman** [comment]
Ok thanks I'll look in to that once I have succeed the brain extraction.
here is my initial image:
![t2initial](https://cloud.githubusercontent.com/assets/16150529/24238004/fe8f4ca2-0fa8-11e7-9aa9-a4cccb5b9c19.png)
as you see the resolution is not isotropic, I don't know if it's a huge problem or not.
and here is the results:
![t2results](https://cloud.githubusercontent.com/assets/16150529/24238024/143f1302-0fa9-11e7-80a3-1687bc355358.png)

is there a way to have a file with the labels grey and white matter ? 

I am trying using different template but without success for now.

**Comment by ntustison** [comment]
Isotropy isn't the big problem---more likely it's the orientation of the head and the lack of superior and inferior background voxels.  You should try padding---``ImageMath ... PadImage ...``.

**Comment by stnava** [comment]
that and the fact that the brain extraction assumes that csf is dark


brian

**Comment by pdeman** [comment]
even when it's a T2 ?

**Comment by stnava** [comment]
the script doesnt know anything about which modality you are using


brian

**Comment by pdeman** [comment]
yes but if it assumes csf is dark it can't work with T2 no ? I though it works with any modalities

**Comment by cookpa** [comment]
The segmentation part does a k-means segmentation (k = 3) and outputs labels in order of intensity. The assumption is T1 intensities, going (from darkest) CSF, GM, WM. Change these lines

https://github.com/stnava/ANTs/blob/master/Scripts/antsBrainExtraction.sh#L496-L498

to make it apply the same steps with a T2 image.

But the segmentation will only work to correct a few voxels around the edges. To do that, you need to have a good registration. Padding the input image will probably help. 

To assess the registration component, run the script with `-k 1` to keep the intermediate output. The image ending in "PriorWarped.nii.gz" shows the mask inferred from the registration step. This should be nearly correct, if it's way off then it means the registration step has failed.

**Comment by stnava** [comment]
good point phil.

worth adding the "light to dark" assumption as an option?

longer term, i've always thought that this step should use priors to define
this.


brian

**Comment by ntustison** [comment]
I've got some time to do this today.  Or do you want to make these changes, Phil?

**Comment by cookpa** [comment]
I'll have a go at generalizing for other modalities

**Comment by pdeman** [comment]
I'll wait before modifying any line then. I'll wait for the option. thanks !

**Comment by ntustison** [comment]
Okay, sounds good.

**Comment by cookpa** [comment]
I just pushed an update that lets the user select tissue classes with `-c`. You specify K, then which classes are CSF, GM, WM respectively. If K > 3, it's assumed the other classes are junk. This worked pretty well for me with K=4 for a T2 image. I checked that I get the same results as before with the defaults, which assume a T1 image and K=3.

https://github.com/stnava/ANTs/commit/646459bf297a50cdf608ea7d39cb79b44906ad22
